geopyspark-backend/geotrellis/src/main/scala/geopyspark/geotrellis/vlm/RasterSource.scala

```diff
@@ -150,6 +256,131 @@ object RasterSource {
+        case null => None
+      }
+
+    readOrderedToLayout(
+      sc,
+      layerType,
+      sc.parallelize(scalaPaths, partitions.getOrElse(sc.defaultParallelism)),
+      layoutType,
+      targetCRS,
+      resampleMethod,
+      readMethod
+    )
+  }
+
+  def readOrderedToLayout(
+    sc: SparkContext,
+    layerType: String,
+    rdd: RDD[Seq[(String, String)]],
+    layoutType: LayoutType,
+    targetCRS: String,
+    resampleMethod: ResampleMethod,
+    readMethod: String
+  ): SpatialTiledRasterLayer = {
+    val rasterSourcesRDD: RDD[Seq[(String, RasterSource)]] =
+      (readMethod match {
+        case GEOTRELLIS =>
+          rdd.mapPartitions { iter =>
+            iter.map { files: Seq[(String, String)] =>
+              files.map { case (k, v) => (k, new GeoTiffRasterSource(v): RasterSource)
+              }
+            }
+          }
+        case GDAL =>
+          rdd.mapPartitions { iter =>
+            iter.map { files: Seq[(String, String)] =>
+              files.map { case (k, v) => (k, GDALRasterSource(v): RasterSource)
+              }
+            }
+          }
+      }).cache()
+
+    val orderedRasterSourcesRDD: RDD[Seq[RasterSource]] =
+      rasterSourcesRDD.map { keyedSource: Seq[(String, RasterSource)] =>
+        keyedSource.sortBy { _._1 }.map { case (_, source) => source }
+      }
+
+    val reprojectedSourcesRDD: RDD[Seq[RasterSource]] =
+      targetCRS match {
+        case crs: String =>
+          orderedRasterSourcesRDD.map {
+            _.map { _.reproject(CRS.fromString(crs), resampleMethod) }
+          }
+        case null =>
+          orderedRasterSourcesRDD
+      }
+
+    val rasterSummary: RasterSummary =
+      reprojectedSourcesRDD
+        .map { sources: Seq[RasterSource] =>
+          sources
+            .map { source =>
+              val ProjectedExtent(extent, crs) = source.getComponent[ProjectedExtent]
+              val cellSize = CellSize(extent, source.cols, source.rows)
+              RasterSummary(crs, source.cellType, cellSize, extent, source.size, 1)
+            }.reduce { _ combine _ }
+        }
+        .reduce { _ combine _ }
+
+    val LayoutLevel(zoom, layout) =
+      layoutType match {
+        case global: GlobalLayout =>
+          val scheme = ZoomedLayoutScheme(rasterSummary.crs, global.tileSize)
+          scheme.levelForZoom(global.zoom)
+        case local: LocalLayout =>
+          val scheme = FloatingLayoutScheme(local.tileCols, local.tileRows)
+          rasterSummary.levelFor(scheme)
+      }
+
+    val tileLayerMetadata: TileLayerMetadata[SpatialKey] =
+      rasterSummary.toTileLayerMetadata(layout, zoom)._1
+
+    val layoutsRDD: RDD[Seq[LayoutTileSource]] =
+      reprojectedSourcesRDD.map { _.map { _.tileToLayout(layout, resampleMethod) } }
+
+    val readInSourcesRDD: RDD[Seq[(SpatialKey, MultibandTile)]] =
+      layoutsRDD.map { layouts: Seq[LayoutTileSource] =>
+        layouts.flatMap { _.readAll() }
```

Comment:
In reference to above comment: at this point all the tiles are actually read into memory in random order, partially sorted per partition. At some point shuffle needed to happen per `SpatialKey` before this happened.